go/vt/servenv/grpc_server_auth_spiffe.go

```diff
@@ -0,0 +1,162 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servenv
+
+import (
+	"context"
+	"crypto/x509"
+	"net/url"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"vitess.io/vitess/go/vt/log"
+)
+
+var (
+	// spiffeTrustDomains list of allowed SPIFFE Trust Domains for client SVIDs during authorization
+	spiffeTrustDomains string
+	// SPIFFEAuthPlugin implements AuthPlugin interface
+	_ Authenticator = (*SPIFFEAuthPlugin)(nil)
+)
+
+// The datatype for spiffe auth Context keys
+type spiffeIdKey int
+
+const (
+	// Internal Context key for the authenticated SPIFFE ID
+	spiffeId spiffeIdKey = 0
+)
+
+func registerGRPCServerAuthSPIFFEFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&spiffeTrustDomains, "grpc_auth_spiffe_allowed_trust_domains", spiffeTrustDomains, "List of allowed SPIFFE Trust Domains for client SVIDs (separated by comma).")
+}
+
+// SPIFFEAuthPlugin implements X.509-based SVID for SPIFFE authentication for grpc. It contains an array of trust domains
+// that will be authorized to connect to the grpc server.
+type SPIFFEAuthPlugin struct {
+	spiffeTrustDomains []string
+}
+
+// Authenticate implements Authenticator interface. This method will be used inside a middleware in grpc_server to authenticate
+// incoming requests.
+func (spa *SPIFFEAuthPlugin) Authenticate(ctx context.Context, fullMethod string) (context.Context, error) {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, status.Errorf(codes.Unauthenticated, "no peer connection info")
+	}
+	tlsInfo, ok := p.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return nil, status.Errorf(codes.Unauthenticated, "not connected via TLS")
+	}
+
+	cert := tlsInfo.State.PeerCertificates[0] // Only check the leaf certificate
+	spiffeIdUrl, ok := validateSVIDCert(cert, spa.spiffeTrustDomains)
+	if !ok {
+		return nil, status.Errorf(codes.Unauthenticated, "client certificate not authorized")
+	}
+
+	log.Infof("SPIFFE auth plugin has authenticated client with SPIFFE ID %v", spiffeId)
+	return newSPIFFEAuthContext(ctx, spiffeIdUrl), nil
+}
+
+// Validates the given certificate as a valid SVID leaf certificate based on
+// https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#4-constraints-and-usage
+// and https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#5-validation
+func validateSVIDCert(cert *x509.Certificate, trustedDomains []string) (*url.URL, bool) {
+	issuedForTrustedDomain := false
+	spiffeId := &url.URL{}
```

Comment:
As a nit: I don't see any reason to pre-allocate this `url.URL` up front, rather than just returning `nil` on the fail paths. Nowhere else down the line seems to expect this to be a concrete object, and given this is on the authenticate path, it'll be rather hot.

I'd prefer to see `return nil, false` elsewhere, then on success, a `return cert.URI[0], true` seems sufficient.